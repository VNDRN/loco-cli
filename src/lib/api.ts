import fetch from "isomorphic-unfetch";
import { PullOptions, PushOptions, Translations } from "../../types";

const fetchApi = async <T>(path: string, opts = {}) => {
  const response = await fetch(`https://localise.biz/api${path}`, opts);
  if (!response.ok) {
    throw new Error(`HTTPError: ${response.status} ${response.statusText}`);
  }
  const json = (await response.json()) as T;
  return json;
};

export const apiPull = (key: string, options: PullOptions = {}) =>
  fetchApi<Translations>(
    `/export/all.json?${new URLSearchParams({ key, ...options }).toString()}`
  );

export const apiPush = (
  key: string,
  locale: string,
  translations: Translations[string],
  options: PushOptions = {}
) =>
  fetchApi<void>(
    `/import/json?
  ${new URLSearchParams({
    key,
    locale,
    ...Object.keys(options).reduce(
      (acc, key) => ({
        ...acc,
        [key]: options[key as keyof PushOptions]?.toString(),
      }),
      {}
    ),
  }).toString()}`,
    {
      method: "POST",
      body: JSON.stringify(translations),
    }
  );
