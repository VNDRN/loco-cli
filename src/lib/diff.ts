import { detailedDiff } from "deep-object-diff";
import { Translations } from "../../types";
import { dotObject } from "./dotObject";

interface DetailedDiff {
  added: object;
  updated: object;
  deleted: object;
}

export const diff = (local: Translations, remote: Translations) => {
  const { added, updated, deleted } = detailedDiff(
    local,
    remote
  ) as DetailedDiff;

  const addedRes = dotObject(added);
  const updatedRes = dotObject(updated);
  const deletedRes = dotObject(deleted);
  return {
    totalCount:
      Object.keys(addedRes).length +
      Object.keys(updatedRes).length +
      Object.keys(deletedRes).length,
    added: addedRes,
    addedCount: Object.keys(addedRes).length,
    updated: updatedRes,
    updatedCount: Object.keys(updatedRes).length,
    deleted: deletedRes,
    deletedCount: Object.keys(deletedRes).length,
  };
};
