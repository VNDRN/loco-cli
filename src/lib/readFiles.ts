import { existsSync, readdirSync, stat, statSync } from "fs";
import { readFile } from "fs";
import { join } from "path";
import { Translations } from "../types";
import { exitError } from "../util/exit";

const readJSON = async (path: string) => {
  if (!existsSync(path)) {
    exitError(`File not found: ${path}`);
  }

  return new Promise((resolve, reject) => {
    readFile(path, "utf8", (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(JSON.parse(data));
    });
  }) as Promise<Record<string, string>>;
};

const readFilesInDir = async (path: string, separator?: string) => {
  const res = {};
  if (!existsSync(path)) {
    exitError(`Directory not found: "${path}"`);
  }
  const files = readdirSync(path);

  await Promise.all(
    files.map(async (file) => {
      if (file.endsWith(".json")) {
        const json = await readJSON(join(path, file));
        Object.keys(json).forEach((key) => {
          // @ts-expect-error
          res[`${file.replace(".json", "")}${separator}${key}`] = json[key];
        });
      }
    })
  );
  return res;
};

export const readFiles = async (
  dir: string,
  useNamespaces: boolean
): Promise<Translations> => {
  if (!existsSync(dir)) {
    throw new Error(`Directory not found: "${dir}"`);
  }
  const translations: Translations = {};
  const locales = readdirSync(dir);
  await Promise.all(
    locales.map(async (locale) => {
      const localeDir = join(dir, locale);
      if (useNamespaces) {
        // using namespaces; locales should be folders
        if (statSync(localeDir).isDirectory()) {
          translations[locale] = await readFilesInDir(localeDir, ":");
        }
      } else {
        // using namespaces; locales shoud be JSON files
        if (locale.endsWith(".json")) {
          translations[locale.replace(".json", "")] = await readJSON(
            join(dir, locale)
          );
        }
      }
    })
  );
  return translations;
};
