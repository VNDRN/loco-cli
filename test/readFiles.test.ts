import { readFiles } from "../src/lib/readFiles";

test("readFiles without namespaces", async () => {
  expect(await readFiles("example/no-namespaces", false)).toMatchSnapshot();
});

test("readFiles with namespaces", async () => {
  expect(await readFiles("example/namespaces", true)).toMatchSnapshot();
});
