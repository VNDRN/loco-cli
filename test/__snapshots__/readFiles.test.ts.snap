// Vitest Snapshot v1

exports[`readFiles with namespaces 1`] = `
{
  "en": {
    "common:actions": {
      "confirm": "Confirm",
      "dismiss": "Dismiss",
    },
    "home:hero": {
      "description": "It is used in the tests",
      "title": "Welcome to the dummy data",
    },
  },
  "es": {
    "common:actions": {
      "confirm": "Confirmar",
      "dismiss": "Despedir",
    },
    "home:hero": {
      "description": "Se utiliza en las pruebas",
      "title": "Bienvenido a los datos ficticios",
    },
  },
  "nl": {
    "common:actions": {
      "confirm": "Bevestigen",
      "dismiss": "Annuleren",
    },
    "home:hero": {
      "description": "Ze wordt gebruikt bij de testen",
      "title": "Welkom in de dummy data",
    },
  },
}
`;

exports[`readFiles without namespaces 1`] = `
{
  "en": {
    "common:actions": {
      "confirm": "Confirm",
      "dismiss": "Dismiss",
    },
    "home:hero": {
      "description": "It is used in the tests",
      "title": "Welcome to the dummy data",
    },
  },
  "es": {
    "common:actions": {
      "confirm": "Confirmar",
      "dismiss": "Despedir",
    },
    "home:hero": {
      "description": "Se utiliza en las pruebas",
      "title": "Bienvenido a los datos ficticios",
    },
  },
  "nl": {
    "common:actions": {
      "confirm": "Bevestigen",
      "dismiss": "Annuleren",
    },
    "home:hero": {
      "description": "Ze wordt gebruikt bij de testen",
      "title": "Welkom in de dummy data",
    },
  },
}
`;
